vslib/src/MACsec.cpp

```diff
@@ -0,0 +1,640 @@
+    if (m_macsecfd < 0)
+    {
+        const std::string error_msg =
+            "failed to open macsec socket " + m_macsec_interface_name + ", errno: " + std::to_string(errno);
+        SWSS_LOG_ERROR("%s", error_msg.c_str());
+        throw std::system_error(errno, std::generic_category(), error_msg);
+    }
+
+    struct sockaddr_ll sock_address;
+    memset(&sock_address, 0, sizeof(sock_address));
+    sock_address.sll_family = PF_PACKET;
+    sock_address.sll_protocol = htons(ETH_P_ALL);
+    sock_address.sll_ifindex = if_nametoindex(m_macsec_interface_name.c_str());
+    if (sock_address.sll_ifindex == 0)
+    {
+        const std::string error_msg =
+            "failed to get interface index for" + m_macsec_interface_name;
+        SWSS_LOG_ERROR("%s", error_msg);
+        close(m_macsecfd);
+        throw std::runtime_error(error_msg);
+    }
+
+    if (SwitchStateBase::promisc(m_macsec_interface_name.c_str()))
+    {
+        const std::string error_msg =
+            "promisc failed on " + m_macsec_interface_name;
+        SWSS_LOG_ERROR("%s", error_msg);
+        close(m_macsecfd);
+        throw std::runtime_error(error_msg);
+    }
+
+    if (bind(
+            m_macsecfd,
+            (struct sockaddr *)&sock_address,
+            sizeof(sock_address)) < 0)
+    {
+        const std::string error_msg =
+            "bind failed on " + m_macsec_interface_name;
+        SWSS_LOG_ERROR("%s", error_msg);
+        close(m_macsecfd);
+        throw std::runtime_error(error_msg);
+    }
+
+    m_forward_thread =
+        std::make_shared<std::thread>(&MACsecForwarder::forward, this);
+
+    SWSS_LOG_NOTICE(
+        "setup MACsec forward rule for %s succeeded",
+        m_macsec_interface_name.c_str());
+}
+
+MACsecForwarder::~MACsecForwarder()
+{
+    SWSS_LOG_ENTER();
+
+    m_run_thread = false;
+    m_exit_event.notify();
+    m_forward_thread->join();
+    int err = close(m_macsecfd);
+    if (err)
+    {
+        SWSS_LOG_ERROR(
+            "failed to remove macsec device: %s, err: %d",
+            m_macsec_interface_name.c_str(),
+            err);
+    }
+}
+
+int MACsecForwarder::get_macsecfd() const
+{
+    SWSS_LOG_ENTER();
+
+    return m_macsecfd;
+}
+
+void MACsecForwarder::forward()
+{
+    SWSS_LOG_ENTER();
+
+    unsigned char buffer[ETH_FRAME_BUFFER_SIZE];
+
+    swss::Select s;
+    SelectableFd fd(m_macsecfd);
+
+    s.addSelectable(&m_exit_event);
+    s.addSelectable(&fd);
+
+    while (m_run_thread)
+    {
+        swss::Selectable *sel = NULL;
+
+        int result = s.select(&sel);
+
+        if (result != swss::Select::OBJECT)
+        {
+            SWSS_LOG_ERROR(
+                "selectable failed: %d, ending thread for %s",
+                result,
+                m_macsec_interface_name.c_str());
+            return;
+        }
+
+        if (sel == &m_exit_event) // thread end event
+            break;
+
+        ssize_t size = read(m_macsecfd, buffer, sizeof(buffer));
```

Comment:
this is blocking function, if there is no data then this will block forever, and there no data flowing, and you will try to close thread, you will have deadlock here

Reply:
I agree your point. It should be fixed. But I have a question whether this bug will also happen at `read tapfd` https://github.com/Azure/sonic-sairedis/blob/master/vslib/src/HostInterfaceInfo.cpp#L254 ?

Reply:
no it will not, i missed the selectable parth here, this mean if we reach this point there will always be something to read